spmd/tensor/device_mesh.py

```diff
@@ -40,6 +40,16 @@ def set_global_device_mesh(mesh: Optional["DeviceMesh"]) -> None:
     _global_device_mesh = mesh
 
 
+T = TypeVar("T")
+_L = Union[T, Sequence[T]]
+DeepList = _L[_L[_L[_L[_L[int]]]]]
```

Comment:
This was failing mypy typechecking on construction.

The type is meant to be "n-dimensional int Tensors, and things that torch.as_tensor(t) will convert cleanly to n-dimensional int tensors".

The later cannot be expressed cleanly in mypy types; so the choice was to remove the annotation (or use Any), or to do this.

This is the standard way to simulate recursive structure types in mypy. It is not as general as true recursive type support, but this will handle up to 5-dimensional expressions.

Reply:
`torch.tensor(mesh, dtype=torch.int)` is literally do the same thing as `torch.as_tensor` by "convert" cleanly, and that does not require to have this pseudo recursive type definition for mypy, what I am wondering is that why we need to switch to `torch.as_tensor`, I don't see a strong reason on this, that's why I think we need more discussion on it. Also this seems not relating to this PR, if you want to change this, feel free to open another PR and we can discuss further

Reply:
the type that DeviceMesh took was _already_ broken. It worked in one very narrow range, that of `[[int]]`; it didn't handle `[int]` or `[[[int]]]`; and it was throwing mypy errors when I used it in testing.

I rewrote the type annotation to fix this, rather than putting `# type: ignore` at the call site.

The constructor cleanup was not needed, I was just already there, and `.as_tensor()` is cleaner than what was being done (and we already weren't doing a clean copy, so I fixed that too).

This change is more correct, and more flexible.

Are you arguing against doing this, or that these cleanups should be in seperate PRs?

Reply:
I think it's good to fix the type annotation, i'm not against that, but there're several options to avoid this DeepList annotations. Also I am curious why it breaks the test because I manually create a 3d int list and mypy didn't complain about it at all. I think It should at least in this separate PR and we can discuss further there, i.e. mypy released their experimental recursive types support https://mypy-lang.blogspot.com/2022/09/mypy-0981-released.html Maybe we want to use this for a cleaner type annotation

Reply:
Lifted out into seperate PR: https://github.com/pytorch/tau/pull/515